hbase-client/src/main/java/org/apache/hadoop/hbase/client/ImmutableScan.java

```diff
@@ -0,0 +1,477 @@
+import java.util.Map;
+import java.util.NavigableSet;
+import org.apache.hadoop.hbase.filter.Filter;
+import org.apache.hadoop.hbase.io.TimeRange;
+import org.apache.hadoop.hbase.security.access.Permission;
+import org.apache.hadoop.hbase.security.visibility.Authorizations;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * Immutable version of Scan
+ */
+@InterfaceAudience.Public
+public final class ImmutableScan extends Scan {
+
+  /**
+   * Create Immutable instance of Scan from given Scan object
+   *
+   * @param scan Copy all values from Scan
+   * @throws IOException From parent constructor
+   */
+  public ImmutableScan(Scan scan) throws IOException {
+    super(scan);
+    super.setIsolationLevel(scan.getIsolationLevel());
+    Map<byte[], NavigableSet<byte[]>> familyMap = scan.getFamilyMap();
+    for (Map.Entry<byte[], NavigableSet<byte[]>> entry : familyMap.entrySet()) {
+      byte[] family = entry.getKey();
+      NavigableSet<byte[]> cols = entry.getValue();
+      if (cols != null && cols.size() > 0) {
+        for (byte[] col : cols) {
+          super.addColumn(family, col);
+        }
+      } else {
+        super.addFamily(family);
+      }
+    }
+    for (Map.Entry<String, byte[]> attr : scan.getAttributesMap().entrySet()) {
+      super.setAttribute(attr.getKey(), attr.getValue());
+    }
+    for (Map.Entry<byte[], TimeRange> entry : scan.getColumnFamilyTimeRange().entrySet()) {
+      TimeRange tr = entry.getValue();
+      super.setColumnFamilyTimeRange(entry.getKey(), tr.getMin(), tr.getMax());
+    }
+    super.setPriority(scan.getPriority());
+  }
+
+  /**
+   * Create Immutable instance of Scan from given Get object
+   *
+   * @param get Get to model Scan after
+   */
+  public ImmutableScan(Get get) {
+    super(get);
+    super.setIsolationLevel(get.getIsolationLevel());
+    for (Map.Entry<String, byte[]> attr : get.getAttributesMap().entrySet()) {
+      super.setAttribute(attr.getKey(), attr.getValue());
+    }
+    for (Map.Entry<byte[], TimeRange> entry : get.getColumnFamilyTimeRange().entrySet()) {
+      TimeRange tr = entry.getValue();
+      super.setColumnFamilyTimeRange(entry.getKey(), tr.getMin(), tr.getMax());
+    }
+    super.setPriority(get.getPriority());
+  }
+
+  /**
+   * Create a new Scan with a cursor. It only set the position information like start row key.
+   * The others (like cfs, stop row, limit) should still be filled in by the user.
+   * {@link Result#isCursor()}
+   * {@link Result#getCursor()}
+   * {@link Cursor}
+   */
+  public static Scan createScanFromCursor(Cursor cursor) {
+    Scan scan = new Scan().withStartRow(cursor.getRow());
+    try {
+      return new ImmutableScan(scan);
+    } catch (IOException e) {
+      throw new RuntimeException("Scan should not throw IOException", e);
+    }
+  }
+
+  @Override
+  public Scan addFamily(byte[] family) {
+    throw new IllegalStateException("ImmutableScan does not allow access to addFamily");
```

Comment:
UnsupportedOperationException or IllegalStateException?

Reply:
Oh, UnsupportedOperationException is better. will update.